src/realm/sync/noinst/client_reset.cpp

```diff
@@ -1799,17 +1803,148 @@ void client_reset::transfer_group(const Transaction& group_src, Transaction& gro
-                                                                      sync::SaltedFileIdent client_file_ident,
-                                                                      util::Logger& logger,
-                                                                      bool recover_local_changes)
+// A table without a "class_" prefix will not generate sync instructions.
+const char* meta_reset_table_name = "client_reset_metadata";
+const char* pk_col_name = "id";
+const char* version_column_name = "version";
+const char* timestamp_col_name = "event_time";
+const char* sequence_col_name = "sequence";
+const char* reset_type_col_name = "type_of_reset";
+constexpr int64_t metadata_version = 1;
+
+void remove_pending_client_resets(TransactionRef wt)
+{
+    REALM_ASSERT(wt);
+    if (auto table = wt->get_table(meta_reset_table_name)) {
+        if (table->size()) {
+            table->clear();
+        }
+    }
+}
+
+util::Optional<PendingReset> has_pending_reset(Transaction& rt)
+{
+    ConstTableRef table = rt.get_table(meta_reset_table_name);
+    if (!table || table->size() == 0) {
+        return util::none;
+    }
+    ColKey timestamp_col = table->get_column_key(timestamp_col_name);
+    ColKey type_col = table->get_column_key(reset_type_col_name);
+    ColKey version_col = table->get_column_key(version_column_name);
+    ColKey sequence_col = table->get_column_key(sequence_col_name);
+    REALM_ASSERT(timestamp_col);
+    REALM_ASSERT(type_col);
+    REALM_ASSERT(version_col);
+    REALM_ASSERT(sequence_col);
+    constexpr bool is_ascending = false;
+    auto sorted = table->get_sorted_view(sequence_col, is_ascending);
+    Obj first = sorted.get_object(0);
+    REALM_ASSERT(first);
+    PendingReset pending;
+    int64_t version = first.get<int64_t>(version_col);
+    pending.time = first.get<Timestamp>(timestamp_col);
+    pending.sequence = first.get<int64_t>(sequence_col);
+    if (version > metadata_version) {
+        throw ClientResetFailed(
+            util::format("Unsupported client reset metadata version: %1 vs %2, from %3, sequence %4", version,
+                         metadata_version, pending.time, pending.sequence));
+    }
+    int64_t type = first.get<int64_t>(type_col);
+    if (type == 0) {
+        pending.type = AutomaticResetType::DiscardLocal;
+    }
+    else if (type == 1) {
+        pending.type = AutomaticResetType::Recover;
+    }
+    else {
+        throw ClientResetFailed(
+            util::format("Unsupported client reset metadata type: %1 from %2", type, pending.time));
+    }
+    return pending;
+}
+
+void track_reset(Transaction& wt, AutomaticResetType mode)
+{
+    TableRef table = wt.get_table(meta_reset_table_name);
+    ColKey version_col, timestamp_col, type_col, sequence_col;
+    if (!table) {
+        table = wt.add_table_with_primary_key(meta_reset_table_name, type_ObjectId, pk_col_name);
```

Comment:
We don't strictly need a pk here, however I think it might make future migrations between versions easier. I have removed the sequence column though due to this and your other comments.